drivers/kscan/kscan_mchp_xec.c

```diff
@@ -0,0 +1,405 @@
+/* Thread stack size */
+#define TASK_STACK_SIZE 1024
+
+struct kscan_xec_data {
+	/* variables in usec units */
+	u32_t deb_time_press;
+	u32_t deb_time_rel;
+	s64_t poll_timeout;
+	u32_t poll_period;
+	u8_t matrix_stable_state[MAX_MATRIX_KEY_COLS];
+	u8_t matrix_unstable_state[MAX_MATRIX_KEY_COLS];
+	u8_t matrix_previous_state[MAX_MATRIX_KEY_COLS];
+	/* Index in to the scan_clock_cycle to indicate start of debouncing */
+	u8_t scan_cycle_idx[MAX_MATRIX_KEY_COLS][MAX_MATRIX_KEY_ROWS];
+	/* Track previous "elapsed clock cycles" per matrix scan. This
+	 * is used to calculate the debouncing time for every key
+	 */
+	u8_t scan_clk_cycle[SCAN_OCURRENCES];
+	struct k_sem poll_lock;
+	u8_t scan_cycles_idx;
+	kscan_callback_t callback_isr;
+	struct k_thread thread;
+	atomic_t enable_scan;
+
+	K_THREAD_STACK_MEMBER(thread_stack, TASK_STACK_SIZE);
+};
+
+static KSCAN_Type *base = (KSCAN_Type *)
+	(DT_INST_0_MICROCHIP_XEC_KSCAN_BASE_ADDRESS);
+
+static struct kscan_xec_data kbd_data;
+
+static void drive_keyboard_column(int data)
+{
+	if (data == KEYBOARD_COLUMN_DRIVE_ALL) {
+		/* KSO output controlled by the KSO_SELECT field */
+		base->KSO_SEL = MCHP_KSCAN_KSO_ALL;
+	} else if (data == KEYBOARD_COLUMN_DRIVE_NONE) {
+		/* Keyboard scan disabled. All KSO output buffers disabled */
+		base->KSO_SEL = MCHP_KSCAN_KSO_EN;
+	} else {
+		/* It is assumed, bit 5 was previously set */
+		base->KSO_SEL = data;
+	}
+}
+
+static u8_t read_keyboard_row(void)
+{
+	/* In this implementation a 1 means key pressed */
+	return ~(base->KSI_IN & 0xFF);
+}
+
+static bool is_matrix_ghosting(const uint8_t *state)
+{
+	/* matrix keyboard designs are suceptible to ghosting.
+	 * An extra key appears to be pressed when 3 keys
+	 * belonging to the same block are pressed.
+	 * for example, in the following block
+	 *
+	 * . . w . q .
+	 * . . . . . .
+	 * . . . . . .
+	 * . . m . a .
+	 *
+	 * the key m would look as pressed if the user pressed keys
+	 * w, q and a simultaneously. A block can also be formed,
+	 * with not adjacent columns.
+	 */
+	for (int c = 0; c <  MAX_MATRIX_KEY_COLS; c++) {
+		if (!state[c])
+			continue;
+
+		for (int c_n = c + 1; c_n <  MAX_MATRIX_KEY_COLS; c_n++) {
+			/* we and the columns to detect a "block".
+			 * this is an indication of ghosting, due to current
+			 * flowing from a key which was never pressed. in our
+			 * case, current flowing is a bit set to 1 as we
+			 * flipped the bits when the matrix was scanned.
+			 * now we or the colums using z&(z-1) which is
+			 * non-zero only if z has more than one bit set.
+			 */
+			u8_t common_row_bits = state[c] & state[c_n];
+
+			if (common_row_bits & (common_row_bits - 1))
+				return true;
+		}
+	}
+
+	return false;
+}
+
+static bool read_keyboard_matrix(u8_t *new_state)
+{
+	u8_t row;
+	u8_t key_event = 0U;
+
+	for (int col = 0; col < MAX_MATRIX_KEY_COLS; col++) {
+		/* Select column, then allow the value to stabilize */
+		drive_keyboard_column(col);
+		k_busy_wait(50U);
```

Comment:
Add comment indicating this time for ramping up

Reply:
done